weave/monitoring2/openai.py

```diff
@@ -0,0 +1,401 @@
+
+Callbacks = List[Callable]
+# Context = Dict[str, Any]
+
+
+class Context(BaseModel):
+    model_config = ConfigDict(extra="allow")
+
+    inputs: Optional[BaseModel] = None
+    outputs: Optional[BaseModel] = None
+
+
+class Callback:
+    def before_send_request(self, context: Context, *args, **kwargs):
+        ...
+
+    def before_end(self, context: Context, *args, **kwargs):
+        ...
+
+    def before_yield_chunk(self, context: Context, *args, **kwargs):
+        ...
+
+    def after_yield_chunk(self, context: Context, *args, **kwargs):
+        ...
+
+
+class InfoMsg(Callback):
+    def before_send_request(self, context: Context, *args, **kwargs):
+        info(f"before send request, {context=}")
+
+    def before_end(self, context: Context, *args, **kwargs):
+        info(f"before end, {context=}")
+
+    def before_yield_chunk(self, context: Context, *args, **kwargs):
+        info(f"before yield chunk, {context=}")
+
+    def after_yield_chunk(self, context: Context, *args, **kwargs):
+        info(f"after yield chunk, {context=}")
+
+
+class ReassembleStream(Callback):
+    def before_end(self, context: Context, *args, **kwargs):
+        if hasattr(context, "chunks"):
+            input_messages = context.inputs.messages
+            context.outputs = reconstruct_completion(input_messages, context.chunks)
+
+
+class LogToStreamTable(Callback):
+    def __init__(self, streamtable: StreamTable):
+        self._streamtable = streamtable
+
+    @classmethod
+    def from_stream_name(cls, stream: str, project: Optional[str] = None, entity: Optional[str] = None):
+        streamtable = StreamTable(stream, project_name=project, entity_name=entity)
+        return cls(streamtable)
+
+    @classmethod
+    def from_stream_key(cls, stream_key: str):
+        tokens = stream_key.split("/")
+        if len(tokens) == 2:
+            project_name, stream_name = tokens
+            entity_name = None
+        elif len(tokens) == 3:
+            entity_name, project_name, stream_name = tokens
+        else:
+            raise ValueError("stream_key must be of the form 'entity/project/stream_name' or 'project/stream_name'")
+
+        streamtable = StreamTable(stream_name, project_name=project_name, entity_name=entity_name)
+        return cls(streamtable)
+
+    def before_send_request(self, context: Context, *args, **kwargs):
+        sig = match_signature(old_create, *args, **kwargs)
+        context.inputs = ChatCompletionRequest.model_validate(sig)
+
+    def before_end(self, context: Context, *args, **kwargs):
+        inputs: ChatCompletionRequest = context.inputs
+        outputs: ChatCompletion = context.outputs
+
+        d = {}
+        if inputs:
+            d["inputs"] = inputs.model_dump()
+        if outputs:
+            d["outputs"] = outputs.model_dump()
+
+        self._streamtable.log(d)
+        self._streamtable.finish()
+
+
+class AsyncChatCompletions:
```

Comment:
Nit: It might not be possible or clean, but i wonder if there is any way to share more code between these two classes